File: backend/app/models/patient.py
"""
`patients` holds both the kiosk-collected identification (mode, hospital reg
number, phone — `PatientIdentification` in the frontend) and the queue-facing
fields the doctor dashboard reads (`Patient` in the frontend). They are kept
as one table because a patient row is created exactly once, at
`POST /kiosk/submissions`, from exactly one `KioskSubmission`.
"""

from __future__ import annotations

import datetime as dt

from sqlalchemy import JSON, DateTime, Enum, Integer, String, Text
from sqlalchemy import true as sa_true
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Patient(Base):
    __tablename__ = "patients"

    # The OPD token doubles as the primary key — the frontend's `Patient.id`
    # and `Patient.token` are always the same value (see patientService.ts).
    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    token: Mapped[str] = mapped_column(String(32), unique=True, index=True)

    name: Mapped[str] = mapped_column(String(120))
    age: Mapped[int] = mapped_column(Integer)
    gender: Mapped[str] = mapped_column(Enum("Male", "Female", "Other", name="gender"))
    abha: Mapped[str] = mapped_column(String(64), default="—")

    # Raw identification as captured at the kiosk (PatientIdentification).
    identification_mode: Mapped[str] = mapped_column(
        Enum("abha", "scan", "new", name="identification_mode"), default="new"
    )
    hospital_reg_number: Mapped[str | None] = mapped_column(String(64), nullable=True)
    phone: Mapped[str | None] = mapped_column(String(32), nullable=True)

    priority: Mapped[str] = mapped_column(Enum("P1", "P2", "P3", name="priority"))
    status: Mapped[str] = mapped_column(
        Enum("Waiting", "In Consultation", "Completed", name="patient_status"),
        default="Waiting",
    )

    complaint: Mapped[str] = mapped_column(Text)
    red_flag: Mapped[bool] = mapped_column(default=False)
    flags: Mapped[list[str]] = mapped_column(JSON, default=list)
    # True once a doctor has explicitly marked the patient reviewed (see
    # `POST /patients/{id}/status` with `resolveRedFlag`). The original
    # `red_flag` / `flags` / `priority` triage result is never overwritten —
    # this only controls whether the queue's "needs urgent review" banner
    # still treats the patient as active.
    red_flag_resolved: Mapped[bool] = mapped_column(default=False)

    submitted_at: Mapped[dt.datetime] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[dt.datetime] = mapped_column(DateTime(timezone=True), default=dt.datetime.utcnow)
    updated_at: Mapped[dt.datetime] = mapped_column(
        DateTime(timezone=True), default=dt.datetime.utcnow, onupdate=dt.datetime.utcnow
    )

    clinical_history: Mapped["ClinicalHistory | None"] = relationship(  # noqa: F821
        back_populates="patient", uselist=False, cascade="all, delete-orphan"
    )
    interviews: Mapped[list["Interview"]] = relationship(  # noqa: F821
        back_populates="patient", cascade="all, delete-orphan"
    )
    documents: Mapped[list["Document"]] = relationship(  # noqa: F821
        back_populates="patient", cascade="all, delete-orphan"
    )
    timeline_events: Mapped[list["TimelineEvent"]] = relationship(  # noqa: F821
        back_populates="patient", cascade="all, delete-orphan"
    )
    alerts: Mapped[list["ClinicalAlert"]] = relationship(  # noqa: F821
        back_populates="patient", cascade="all, delete-orphan"
    )
    vitals: Mapped[list["VitalObservation"]] = relationship(  # noqa: F821
        back_populates="patient", cascade="all, delete-orphan"
    )


class Doctor(Base):
    """A clinician who can log in to the dashboard.

    Doubles as the auth/identity record (Phase 1 auth): `email` + `password_hash`
    back JWT login, `role` drives RBAC (currently only "doctor"; "nurse"/"admin"
    can be added later without a schema change since role is a plain string), and
    `is_active` lets an account be disabled. Auth columns are nullable so rows
    created before Phase 1 remain valid until seeded with credentials.
    """

    __tablename__ = "doctors"

    id: Mapped[str] = mapped_column(String(32), primary_key=True)
    name: Mapped[str] = mapped_column(String(120))
    specialty: Mapped[str] = mapped_column(String(120))
    room: Mapped[str] = mapped_column(String(32))
    initials: Mapped[str] = mapped_column(String(4))

    # --- Authentication / RBAC ----------------------------------------------
    email: Mapped[str | None] = mapped_column(String(255), unique=True, index=True, nullable=True)
    password_hash: Mapped[str | None] = mapped_column(String(255), nullable=True)
    role: Mapped[str] = mapped_column(String(20), default="doctor", server_default="doctor")
    is_active: Mapped[bool] = mapped_column(default=True, server_default=sa_true())
    created_at: Mapped[dt.datetime] = mapped_column(
        DateTime(timezone=True), default=dt.datetime.utcnow
    )
    updated_at: Mapped[dt.datetime] = mapped_column(
        DateTime(timezone=True), default=dt.datetime.utcnow, onupdate=dt.datetime.utcnow
    )

File: src/types/patient.ts
import type { IsoDate } from "./common"

export type Gender = "Male" | "Female" | "Other"

/** Triage priority. P1 is most urgent. Assigned by demo rules, never by diagnosis. */
export type Priority = "P1" | "P2" | "P3"

export type PatientStatus = "Waiting" | "In Consultation" | "Completed"

/** How the patient identified themselves at the kiosk. */
export type IdentificationMode = "abha" | "scan" | "new"

export interface PatientIdentification {
  mode: IdentificationMode
  abhaId: string
  hospitalRegNumber: string
  fullName: string
  age: string
  gender: string
  phone: string
}

export interface Patient {
  id: string
  name: string
  age: number
  gender: Gender
  /** OPD token issued at the kiosk. */
  token: string
  /** Ayushman Bharat Health Account ID. Demo values only. */
  abha: string
  priority: Priority
  status: PatientStatus
  /** Chief complaint in the patient's own words, as captured at the kiosk. */
  complaint: string
  /** Human-readable wait time, e.g. "8 min", or "—" when not waiting. */
  waitTime: string
  redFlag: boolean
  /** Short triage flag labels shown as badges in the queue. */
  flags: string[]
  /**
   * True once a doctor has explicitly clicked "Mark as Reviewed". Never set
   * by AI. The original `priority`/`redFlag`/`flags` triage result is never
   * changed by this — it only stops the queue from still treating the
   * patient as an active, unreviewed red-flag case.
   */
  redFlagResolved: boolean
  /** When the kiosk submission was received. */
  submittedAt: IsoDate
}

export interface Doctor {
  id: string
  name: string
  specialty: string
  /** Consultation room, e.g. "OPD 3". */
  room: string
  /** Two-letter avatar fallback. */
  initials: string
}

export type QueueFilter = "all" | "waiting" | "completed"
